packages/wallet/src/redux/protocols/ledger-top-up/reducer.ts

```diff
@@ -0,0 +1,288 @@
+      const newDestination = [oldDestination[0], oldDestination[1]];
+      const ourCommitment = proposeNewConsensus(theirCommitment, newAllocation, newDestination);
+      const signResult = signAndStore(sharedData, ourCommitment);
+      if (!signResult.isSuccess) {
+        return { protocolState: newProtocolState, sharedData };
+      }
+      sharedData = signResult.store;
+
+      const messageRelay = sendCommitmentReceived(
+        theirAddress(channel),
+        protocolState.processId,
+        signResult.signedCommitment.commitment,
+        signResult.signedCommitment.signature,
+        LEDGER_TOP_UP_PROTOCOL_LOCATOR,
+      );
+      sharedData = queueMessage(sharedData, messageRelay);
+    }
+    return { protocolState: states.waitForPostTopUpLedgerUpdate(protocolState), sharedData };
+  }
+
+  return { protocolState, sharedData };
+};
+
+const waitForPostTopUpLedgerUpdateReducer: ProtocolReducer<states.LedgerTopUpState> = (
+  protocolState: states.WaitForPostTopUpLedgerUpdate,
+  sharedData: SharedData,
+  action: LedgerTopUpAction,
+): ProtocolStateWithSharedData<states.LedgerTopUpState> => {
+  if (action.type !== 'WALLET.COMMON.COMMITMENT_RECEIVED') {
+    console.warn(
+      `Ledger Top Up Protocol expected COMMITMENT_RECEIVED received ${action.type} instead.`,
+    );
+    return { protocolState, sharedData };
+  }
+
+  const checkResult = checkAndStore(sharedData, action.signedCommitment);
+
+  if (!checkResult.isSuccess) {
+    return {
+      protocolState: states.failure({ reason: 'Received Invalid Commitment' }),
+      sharedData,
+    };
+  }
+  sharedData = checkResult.store;
+  const isFirstPlayer = helpers.isFirstPlayer(protocolState.ledgerId, sharedData);
+  // Accept consensus if player B
+  if (!isFirstPlayer) {
+    const ourCommitment = acceptConsensus(action.signedCommitment.commitment);
+    const signResult = signAndStore(sharedData, ourCommitment);
+    if (!signResult.isSuccess) {
+      return { protocolState: states.failure({ reason: 'Signature Failure' }), sharedData };
+    }
+    sharedData = signResult.store;
+
+    const messageRelay = sendCommitmentReceived(
+      getOpponentAddress(sharedData, protocolState.ledgerId),
+      protocolState.processId,
+      signResult.signedCommitment.commitment,
+      signResult.signedCommitment.signature,
+      LEDGER_TOP_UP_PROTOCOL_LOCATOR,
+    );
+    sharedData = queueMessage(sharedData, messageRelay);
+  }
+
+  return { protocolState: states.success({}), sharedData };
+};
+
+function calculateTotalTopUp(allocation: string[]): string {
+  return bigNumberify(allocation[2])
+    .add(allocation[3])
+    .toHexString();
+}
+function getOpponentAddress(sharedData: SharedData, channelId: string) {
+  const channel = getChannel(sharedData, channelId);
+  if (!channel) {
+    throw new Error(`Could not find channel for id ${channelId}`);
+  }
+
+  return theirAddress(channel);
+}
+function getLatestCommitment(sharedData: SharedData, channelId: string) {
```

Comment:
This function doesn't seem to belong in the ledger-top-up reducer.

Reply:
It's just a local helper function but it probably does make sense to share it in `reducer-helpers`